File: MAP/C# - Toy language/Lab10/Toy Language/Toy Language/controller/Controller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Toy_Language.model.statement;
using Toy_Language.repository;
using Toy_Language.utils;

namespace Toy_Language.controller
{
    public class Controller
    {
        private IRepository repo;

        public Controller(
            IRepository _repo
            )
        {
            repo = _repo;
        }

        public void AddPrgState(PrgState prgState)
        {
            repo.AddPrgState(prgState);
        }

        private PrgState OneStep(PrgState state)
        {
            IExeStack<IStmt> stk = state.ExeStack;
            if (stk.IsEmpty())
            {
                throw new Exception("Empty stack");
            }

            IStmt crtStmt = stk.Pop();
            return crtStmt.Execute(state);
        }

        public void AllStep()
        {
            PrgState prg = repo.GetCurrentPrg();
            try
            {
                while(true)
                {
                    OneStep(prg);
                    repo.LogPrgStateExec();
                    System.Console.WriteLine(prg.ToString());
                }
            }
            catch (System.Exception ex)
            {
                System.Console.WriteLine();
                System.Console.WriteLine(ex);	
            }
        }
    }
}

File: Year 2/Semester 1/MAP/JavaRepo - Toy language/TestExample/src/controller/Controller.java
package controller;

import model.Exp;
import repository.IRepository;
import repository.Repository;

public class Controller {
    private Repository repo;

    public Controller(Repository _repo){
        repo = _repo;
    }

    public void addExp(Exp exp){
        repo.addExp(exp);
    }

    public void executeOne(Exp exp){
        System.out.println(exp.toString() + "=" + exp.eval());
    }

    public void executeAll(){
        while(this.repo.notFull()){
            executeOne(repo.getExp());
        }
    }
}
